papers/yellow/main.tex

```diff
@@ -748,19 +727,71 @@ \subsection{Signing A Transaction}
 
 
 
+\subsection{Transaction Validation Rules}
+\label{tx-validation}
 
-\subsection{Verifying A Transaction}
+We split validation of a single transaction into two stages. There are $stateless checks$ which could be done by the
+transaction only being presented. Stateful checks are requiring knowledge of the current state, in the form of the whole
+UTXO set or a part of it, namely concrete boxes a transaction is destroying along with proof of authenticity for them~(
+against a root hash included into a last block header).
 
+Stateless checks are:
 
+\begin{itemize}
+    \item{} a transaction must spend at a least one input and create at least one output. A transaction spends no more
+    than $32767$ inputs and creates no more than $32767$ outputs.
+    \item{} all the output amounts must be non-negative.
+    \item{} an output can not contain more than 4 assets. All the assets amounts must be positive.
+    \item{} transaction outputs collectively could not contain more than 16 assets.
+\end{itemize}
+
+\knote{Check box and transaction sizes.}
+
+Stateful checks are:
+
+\begin{itemize}
+    \item{} total input and output ergo amounts must be the same. Adding up should be done with overflow checks.
+    \item{} all the inputs have valid spending proofs. Total cost to verify all the spending proofs should be no more
+            than a limit per block~(adjustable via miners voting).
+    \item{} inputs should have no more than 16 assets collectively. An input can not contain more than 4 assets.
+    \item{} for each kind of asset in the outputs, total output amount for the asset should be no more than the total
+            input amount, if the asset identifier is not equals to identifier of the first input; otherwise, the total
+            output amount must be positive. The latter case corresponds to issuance of a new asset, while the former
+            sets asset preservation rule. Please note that for total input amount of an asset we do not require for
+            strict equality of the input and output amounts: the output amount could be less than input amount~(or zero).
+\end{itemize}
 
 \subsection{Unified Transactions}
 
+\subsection{Full-Block Validation Rules}
+
+Below are rules for block validation when a node is verifying transactions ($VerifyTransactions = 1$)
+
+\begin{itemize}
+    \item{} every transaction in a block is referencing to inputs from the UTXO set or created by previous transactions

```

Comment:
This description allows double spending - 2 transactions that reference to the same input from UTXO.

Reply:
@catena2w double spending should be prevented during UTXO set tree update (or checking UTXO set transformation proof). But currently double-spending prevention is actually broken, see https://github.com/ergoplatform/ergo/issues/387